File: 0322-coin-change/0322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    let cache = {};
    const MAX =Number.MAX_VALUE;
    
    let ans = findMinCoin(amount);
    return ans === MAX ? -1 : ans;
    
    function findMinCoin(amount) {
        if( cache[amount] )
            return cache[amount];
        
        if( amount === 0 )
            return 0
                
        let res = MAX;
        for( let coin of coins ) {
            if( coin <= amount )
                res = Math.min(res, 1 + findMinCoin(amount-coin));
        }
        
        cache[amount] = res;
        return res
    }
};

/*
551


*/

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        INF = float('inf')

        dp = [INF] * (amount + 1)
        dp[0] = 0

        for amount_val in range(1, amount + 1): 
            for coin in coins:
                if coin <= amount_val:
                    dp[amount_val] = min(
                        dp[amount_val], 
                        1 + dp[amount_val - coin]
                    )
                    
        
        return dp[amount] if dp[amount] != INF else -1